infra/Pos/Communication/Relay/Types.hs

```diff
@@ -44,10 +45,17 @@ instance Buildable PropagationMsg where
     build (DataOnlyPM conts) =
         Buildable.build conts
 
--- | Queue of InvMsges which should be propagated.
-type RelayPropagationQueue = TBQueue (Microsecond, PropagationMsg)
+data RelayContext m = RelayContext
```

Comment:
> But this is unacceptable, since the implementation comes from time-warp.

Time-warp exports some interface. It's not a problem even with approach where `m` is a parameter of data type exposed by time-warp.
For example, time-warp may have:
```
myRelay :: MonadMockable m => MyData -> RelayFunctions m
```
Then in `cardano-sl` we can store `MyData` inside `RelayContext` (we obviously need to know about this type, otherwise we can't create `RelayFunctions`, so it's a part of the interface) and we can implement `MonadRelay` by reading `MyData` from the reader context, using `myRelay` and then taking the required function from `RelayFunctions`:

```
  rlyEnqueue :: (Ether.MonadReader' RelayContext m, MonadMockable m) => PropagationMsg -> Maybe NodeId -> m ()
  rlyEnqueue msg nodeId = do
    RelayContext myData <- Ether.ask'
    let RelayFunctions {..} = myRelay myData
    relayFunctionsEnqueue msg nodeId
```

It's a simple transformation from the approach used in time-warp to the approach used in cardano-sl.